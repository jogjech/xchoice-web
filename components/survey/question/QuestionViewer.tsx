import React, { FunctionComponent, useState } from "react";
import { Radio } from "antd";
import { Question } from "../../../models/survey";
import { RadioChangeEvent } from "antd/lib/radio";
import styles from "./QuestionViewer.module.css";

interface Props {
  question: Question;
  selection: number;
  handleSelectionChange: (selection: number) => void;
  submitWithNoSelection: boolean;
}

const QuestionViewer: FunctionComponent<Props> = ({
  question,
  handleSelectionChange,
  submitWithNoSelection,
  selection: inputSelection,
}) => {
  const [selection, setSelection] = useState<number>(inputSelection);

  const onChange = (event: RadioChangeEvent) => {
    setSelection(event.target.value);
    handleSelectionChange(event.target.value);
  };

  return (
    <>
      <div className={styles.questionContainer}>
        <p className={styles.questionTitle}>{question.questionTitle}</p>
        <div>
          <Radio.Group onChange={onChange} value={selection}>
            {question.choices.map((choice, index) => (
              <Radio className={styles.radioStyle} value={index} key={index}>
                {choice.text}
              </Radio>
            ))}
          </Radio.Group>
        </div>
        {submitWithNoSelection && (
          <div className={styles.choiceErrorMessage}>
            No selection detected. Please select one of the choices.
          </div>
        )}
      </div>
    </>
  );
};

export default QuestionViewer;
