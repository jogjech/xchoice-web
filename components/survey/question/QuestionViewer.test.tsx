import React from "react";
import { shallow, ShallowWrapper } from "enzyme";
import QuestionViewer from "./QuestionViewer";

describe("Question viewer", () => {
  let wrapper: ShallowWrapper;
  const question = {
    questionTitle: "Question title",
    choices: [
      {
        text: "Choice A",
      },
      {
        text: "Choice B",
      },
    ],
  };
  const selection = 1;
  const mockHandleSelectionChange = jest.fn();

  beforeEach(() => {
    wrapper = shallow(
      <QuestionViewer
        question={question}
        selection={selection}
        handleSelectionChange={mockHandleSelectionChange}
        submitWithNoSelection={false}
      ></QuestionViewer>
    );
  });

  it("renders correctly", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("contains question title and choices", () => {
    const title = wrapper.find(".questionTitle");
    expect(title.text()).toEqual(question.questionTitle);

    const choices = wrapper.find("Radio");
    expect(choices.length).toBe(question.choices.length);
    choices.forEach((choice, index) => {
      expect(choice.text()).toBe(question.choices[index].text);
    });
  });
});
