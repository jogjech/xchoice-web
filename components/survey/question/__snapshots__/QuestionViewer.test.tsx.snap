// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Question viewer renders correctly 1`] = `
<Fragment>
  <div
    className="questionContainer"
  >
    <p
      className="questionTitle"
    >
      Question title
    </p>
    <div>
      <Memo([object Object])
        onChange={[Function]}
        value={1}
      >
        <Radio
          className="radioStyle"
          key="0"
          type="radio"
          value={0}
        >
          Choice A
        </Radio>
        <Radio
          className="radioStyle"
          key="1"
          type="radio"
          value={1}
        >
          Choice B
        </Radio>
      </Memo([object Object])>
    </div>
  </div>
</Fragment>
`;
